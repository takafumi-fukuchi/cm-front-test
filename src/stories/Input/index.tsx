import { clsx } from "clsx";
import { forwardRef, type ComponentPropsWithRef } from "react";
import styles from "./styles.module.scss";

interface InputProps extends ComponentPropsWithRef<"input"> {
  isInvalid?: boolean;
  required?: boolean;
  label?: string;
  errorMessage?: string;
  showCharacterCount?: boolean;
  inputValue?: number;
  maxLength?: number;
}

export const Input = forwardRef<HTMLInputElement, InputProps>(
  ({ isInvalid, required, label, errorMessage, showCharacterCount, inputValue, maxLength, ...rest }, ref) => {
    const inputClasses = clsx(styles.input, {
      [styles.isInvalid]: isInvalid,
    });

    return (
      <div className={styles.wrapper}>
        <div className={styles.labelAndCountContainer}>
          {label && (
            <label className={styles.label}>
              {label}
              {required && <span className={styles.requiredApostrophe}>*</span>}
            </label>
          )}{" "}
          {showCharacterCount && <div className={styles.characterCount}>{`${inputValue} / ${maxLength}`}</div>}
        </div>
        <input ref={ref} className={inputClasses} {...rest} />
        {isInvalid && errorMessage && <div className={styles.error}>{errorMessage}</div>}
      </div>
    );
  }
);
Input.displayName = "Input";
