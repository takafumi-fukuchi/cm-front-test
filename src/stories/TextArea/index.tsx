import { clsx } from "clsx";
import { forwardRef, useEffect, useState, type ComponentPropsWithRef } from "react";
import styles from "./styles.module.scss";

interface TextAreaProps extends ComponentPropsWithRef<"textarea"> {
  isInvalid?: boolean;
  required?: boolean;
  errorMessage?: string;
  label?: string;
  maxLength?: number; // 最大文字数を指定するためのプロパティ
}

export const TextArea = forwardRef<HTMLTextAreaElement, TextAreaProps>(
  ({ isInvalid, errorMessage, label, required, maxLength, ...rest }, ref) => {
    const [characterCount, setCharacterCount] = useState(0);

    const handleInputChange = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
      setCharacterCount(event.target.value.length);
      if (rest.onChange) {
        rest.onChange(event);
      }
    };

    useEffect(() => {
      if (ref && "current" in ref && ref.current) {
        setCharacterCount(ref.current.value.length);
      }
    }, [ref]);

    const textAreaClasses = clsx(styles.textArea, {
      [styles.isInvalid]: isInvalid,
    });

    return (
      <div className={styles.wrapper}>
        <div className={styles.labelAndCountContainer}>
          {label && (
            <label className={styles.label}>
              {label}
              {required && <span className={styles.requiredApostrophe}>*</span>}
            </label>
          )}{" "}
          {maxLength && (
            <div className={styles.characterCount}>
              {characterCount}/{maxLength}
            </div>
          )}
        </div>
        <textarea ref={ref} className={textAreaClasses} onChange={handleInputChange} maxLength={maxLength} {...rest} />
        {isInvalid && errorMessage && <div className={styles.error}>{errorMessage}</div>}
      </div>
    );
  }
);

TextArea.displayName = "TextArea";
