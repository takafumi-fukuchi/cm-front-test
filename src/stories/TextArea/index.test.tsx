import { composeStories } from "@storybook/react";
import { render } from "@testing-library/react";
import { createRef } from "react";
import * as stories from "./index.stories";
import { TextArea } from ".";

const testCases = Object.values(composeStories(stories)).map((Story) => [Story.storyName, Story]);

describe("TextArea", () => {
  describe("snapshotテスト", () => {
    test.each(testCases)("%s storyのレンダリング結果", async (_storyName, Story) => {
      const tree = await render(<Story />);
      expect(tree.baseElement).toMatchSnapshot();
    });
  });

  test("refをTextAreaコンポーネントに渡すことができる", () => {
    const ref = createRef<HTMLTextAreaElement>();
    render(<TextArea ref={ref} />);
    expect(ref.current).toBeInTheDocument();
  });
});
