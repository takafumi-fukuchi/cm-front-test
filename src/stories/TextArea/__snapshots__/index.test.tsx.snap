// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea snapshotテスト Invalid storyのレンダリング結果 1`] = `
<body>
  <div>
    <textarea
      class="textArea isInvalid"
      placeholder="記事の説明を入力してください"
    />
  </div>
</body>
`;

exports[`TextArea snapshotテスト Sample storyのレンダリング結果 1`] = `
<body>
  <div>
    <textarea
      class="textArea"
      placeholder="記事の説明を入力してください"
    />
  </div>
</body>
`;
